File: functions/liboqs/variants/go/0_fpr_half_7.go
package main

func fpr_half(x uint64) uint64 {
    if ((x >> 52) & 0x7FF) == 0 {
        return x >> 1
    }
    maskedX := x & ((1 << 52) - 1)
    exponent := ((x >> 52) - 1) << 52
    return maskedX | exponent
}


func main(){
    fpr_half(0)
}

File: functions/liboqs/variants/c/0_fpr_half_7.c
#include <stdint.h>

static uint64_t fpr_half(uint64_t x) {
    uint64_t exp = (x >> 52) & 0x7FF;
    if (exp == 0) {
        return x;
    }
    x -= 1ULL << 52;
    exp = (x >> 52) & 0x7FF;
    if (exp == 0) {
        return 0;
    }
    return x;
}


int main(){
    fpr_half(0);
}
